src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,683 @@
+						 st_rodata_size, SYS_MM_MEM_PERM_EXEC);
+	if (ret < 0) {
+		ret = -ENOMEM;
+		goto err;
+	}
+	/* There are modules marked as lib_code. This is code shared between several modules inside
+	 * the library. Load all lib_code modules with first none lib_code module load.
+	 */
+	if (!mod->type.lib_code)
+		ext_lib->mods_exec_load_cnt++;
+
+	if (ext_lib->mods_exec_load_cnt == 1) {
+		struct sof_man_module *module_entry =
+		    (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(0));
+		for (size_t idx = 0; idx < desc->header.num_module_entries;
+				++idx, ++module_entry) {
+			if (module_entry->type.lib_code) {
+				ret = lib_manager_load_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
+							      idx, mod, desc);
+				if (ret < 0)
+					goto err;
+			}
+		}
+	}
+
+	return 0;
+
+err:
+	sys_mm_drv_unmap_region(va_base_text, st_text_size);
+	sys_mm_drv_unmap_region(va_base_rodata, st_rodata_size);
+
+	return ret;
+}
+
+static int lib_manager_unload_module(uint32_t module_id, struct sof_man_module *mod,
+				     struct sof_man_fw_desc *desc)
+{
+	struct ext_library *ext_lib = ext_lib_get();
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	void *va_base_text = (void *)mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
+	size_t st_text_size = mod->segment[SOF_MAN_SEGMENT_TEXT].flags.r.length;
+	void *va_base_rodata = (void *)mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
+	size_t st_rodata_size = mod->segment[SOF_MAN_SEGMENT_RODATA].flags.r.length;
+	int ret_tmp, ret = 0;
+
+	st_text_size = st_text_size * CONFIG_MM_DRV_PAGE_SIZE;
+	st_rodata_size = st_rodata_size * CONFIG_MM_DRV_PAGE_SIZE;
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_text, st_text_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
+
+	dcache_invalidate_region(va_base_text, st_text_size);
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_rodata, st_rodata_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
+
+	dcache_invalidate_region(va_base_rodata, st_rodata_size);
+
+	/* There are modules marked as lib_code. This is code shared between several modules inside
+	 * the library. Unload all lib_code modules with last none lib_code module unload.
+	 */
+	if (!mod->type.lib_code && ext_lib->mods_exec_load_cnt > 0)
+		ext_lib->mods_exec_load_cnt--;
+
+	if (ext_lib->mods_exec_load_cnt == 0) {
+		struct sof_man_module *module_entry =
+		    (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(0));
+		for (size_t idx = 0; idx < desc->header.num_module_entries;
+				++idx, ++module_entry) {
+			if (module_entry->type.lib_code) {
+				ret_tmp =
+				lib_manager_unload_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
+							  idx, mod, desc);
+				if (ret_tmp < 0)
+					ret = ret_tmp;
+			}
+		}
+	}
+
+	return ret;
+}
+
+static uint8_t *lib_manager_get_instance_bss_address(uint32_t module_id, uint32_t instance_id,
+						     struct sof_man_module *mod)
+{
+	uint32_t instance_bss_size =
+		 mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count;
```

Comment:
is the r.length == bss_size * max_count ?  we can get bbs_size from module_config[module_cfg_offset].is_bytes and the bss_size depends on stream format

Reply:
What about case when loadable library is being reconfigured after load? This mechanism is based on flow proved in closed source and works even if we change f.e. format using ipc set config.